File: alignment/c/smith.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
using namespace std;

//scoring scheme
const int d = -1;
const int match = +2;
const int transition = -1;
const int transversion = -2;

int ***grid;

int scoring_matrix[4][4] = 
    {
        {match, transition, transversion, transversion},//C 
        {transition, match, transversion, transversion},//T
        {transversion, transversion, match, transition},//A
        {transversion, transversion, transition, match} //G
    };

int char_to_int(char base){
    if(base == 'C'){
        return 0;
    }
    if(base == 'T'){
        return 1;
    }
    if(base == 'A'){
        return 2;
    }
    else{
        return 3;
    }
}

int s(string x, string y, int i, int j){
    return scoring_matrix[char_to_int(x[i-1])][char_to_int(y[j-1])];
}
pair<int, int> max_zero(int x, int y, int z){
    int max = 0;
    int i = 3;
    if (x > max){
        max = x;
        i = 0;
    }
    if (y > max){
        max = y;
        i = 1;
    }
    if(z > max){
        max = z;
        i = 2;
    }
    return make_pair(i, max);
}
pair<int, int> F(string x, string y, int i, int j){
    if(i == 0 and j == 0){
        return make_pair(3,0);
    }
    if(i > 0 and j > 0){
        pair<int, int> p = max_zero(grid[i-1][j-1][1] + s(x, y, i, j), grid[i-1][j][1] + d, grid[i][j-1][1] + d);
        return p;
    }
    if(i > 0){
        int value = grid[i-1][j][1] + d;
        if(value <= 0){
            return make_pair(3, 0);
        }
        return make_pair(1, value);
    }else{
        int value = grid[i][j-1][1] + d;
        if(value <= 0){
            return make_pair(3, 0);
        }
        return make_pair(2,value);
    }
}

int main(){
    const string x = "GAATTCCGTTA";
    const string y = "GGATCGA";
//    int grid[x.length()][y.length()][2];
    grid = new int**[x.length()+1];
    for(int i = 0; i < x.length() + 1; i++){
        grid[i] = new int*[y.length()+1];
        for(int j = 0; j< y.length() +1; j++){
            grid[i][j] = new int[2];
        }
    }

    for(int i=0; i<= x.length(); i++){
        for(int j=0; j<= y.length(); j++){
            pair<int, int> p = F(x, y, i, j);
            grid[i][j][0] = p.first;
            grid[i][j][1] = p.second;
        }
    }
    int max_i, max_j = 0; 
    int max_value = 0;
    for(int i =0; i <= x.length(); i++){
        for(int j =0; j<= y.length(); j++){
            int val = grid[i][j][1];
            if(val > max_value){
                max_value = val;
                max_i = i;
                max_j = j;
            }
        }
    }
    cout << "score: " << max_value << endl;
    int i = max_i;
    int j = max_j;
    stack<int> arrows;
    while(i > 0 or j > 0){
        int index = grid[i][j][0];
        arrows.push(index);
        if(index == 3){
            break;
        }
        if(index == 0){
            i--;
            j--;
        }else if(index == 1){
            i--;
        }else if(index == 2){
            j--;
        }
    }
    cout << "x y" << endl;
    while(arrows.size() > 0){
        int a = arrows.top();
        arrows.pop();
        if(a == 0){
            cout << x[i] << " " << y[j] << endl;
            i++;
            j++;
        }else if(a == 1){
            cout << x[i] << " " << "-" << endl;
            i++;
        }else if(a == 2){
            cout << "-" << " " << y[j] << endl;
            j++;
        }
    }
    for(int i = 0; i < x.length() + 1; i++){
        for(int j = 0; j< y.length() +1; j++){
            delete[] grid[i][j];
        }
        delete [] grid[i];
    }
    
    return 0;
}

File: alignment/smith.py
import numpy as np
import operator
x = 'ACACACTAGAGCTTTA' 
y = 'AGCACACATTAACTAC' 
d = -1
match = +2
transition = -1
transversion = -2
purines = ('A', 'G')
pyrimidines = ('T', 'C')
grid = np.arange((len(x)+1)*(len(y)+1)*2).reshape(len(x)+1, len(y)+1, 2)

def s(i, j):
    if x[i-1] == y[j-1]:
        return match
    else:
        if (x[i-1] in purines and y[j-1] in purines) or (x[i-1] in pyrimidines and y[j-1] in pyrimidines):
            return transition
        else:
            return transversion

def F(i, j):
    if (i, j) == (0, 0):
        grid[i,j] = [0,0]
        return 0
    else:
        if i > 0 and j > 0:
            index, value = max(enumerate([grid[i - 1, j - 1][0] + s(i, j), grid[i-1, j][0] + d, grid[i, j -1][0] + d, 0]), key=operator.itemgetter(1))
            grid[i,j] = [value,index]
            return value
        if i > 0:
            value = grid[i-1, j][0] + d
            if value <= 0:
                grid[i, j] = [0, 3]
            else:
                grid[i,j] = [value,1]
            return value 
        else:
            value = grid[i, j - 1][0] + d
            if value <= 0:
                grid[i, j] = [0, 3]
            else:
                grid[i, j] = [value,2]
            return value 
for i in range(len(x)+1):
    for j in range(len(y)+1):
        F(i,j)

arrows = []
max_coords = (0,0)
max_value = 0

for i in range(len(x)+1):
    for j in range(len(y)+1):
        if grid[i,j][0] > max_value:
            max_value = grid[i, j][0]
            max_coords = (i,j)
print "score:", max_value
(i, j) = max_coords

while((i > 0 or j > 0) and grid[i, j][1] != 3):
    arrows.append(grid[i,j][1])
    if grid[i,j][1] == 0:
        i -= 1
        j -= 1
    elif grid[i,j][1] == 1:
        i -= 1
    elif grid[i,j][1] == 2:
        j -= 1
xout = []
yout = []

for arrow in reversed(arrows):
    if arrow == 0:
        xout.append(x[i])
        yout.append(y[j])
        i += 1
        j += 1
    elif arrow == 2:
        xout.append('-')
        yout.append(y[j])
        j += 1
    elif arrow == 1:
        xout.append(x[i])
        yout.append('-')
        i += 1

for x in xout:
    print x,
print " "
for y in yout:
    print y,
print " "
